packages/client/src/links/wsLink/wsClient/reconnectManager.ts

```diff
@@ -0,0 +1,94 @@
+import { sleep } from '@trpc/server/unstable-core-do-not-import';
+import { TRPCWebSocketClosedError } from './utils';
+import type { WsConnection } from './wsConnection';
+
+/**
+ * Represents a fatal WebSocket error that prevents reconnection attempts.
+ * This error is used when retrying the connection would consistently fail due to
+ * a non-recoverable issue.
+ */
+export class TRPCWebSocketReconnectFatal extends Error {
+  constructor(opts: { message: string; cause?: unknown }) {
+    super(opts.message, {
+      cause: opts.cause,
+    });
+    this.name = 'TRPCWebSocketReconnectFatal';
+    Object.setPrototypeOf(this, TRPCWebSocketReconnectFatal.prototype);
+  }
+}
+
+/**
+ * Manages WebSocket reconnections with configurable retry logic.
+ * Handles the lifecycle of a WebSocket connection, including error handling and retries.
+ */
+export class ReconnectManager {
+  private connection: WsConnection | undefined;
+
+  constructor(
+    private readonly retryDelayMs: (attemptIndex: number) => number,
+    private readonly callbacks: {
+      onError: (error: TRPCWebSocketClosedError) => void;
+    },
+  ) {}
+
+  public attach(connection: NonNullable<typeof this.connection>) {
+    this.connection = connection;
+    this.connection.wsObservable.subscribe({
+      next: (ws) => {
+        if (!ws) return;
+
+        ws.addEventListener('open', () => {
+          ws.addEventListener('close', (event) => {
+            this.callbacks.onError(
+              new TRPCWebSocketClosedError({
+                message: 'WebSocket closed',
+                cause: event,
+              }),
+            );
+            void this.reconnect();
```

Comment:
please avoid using `void X` as it  means you're missing to handling potential promise rejections

Reply:
Removed all void promise occurrences. I now explicitly use catch(() => null). Additionally, I refactored some async methods to be synchronous when I determined there were no pending resources to wait for.